core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

```diff
@@ -500,14 +523,35 @@ private[spark] object SparkHadoopUtil extends Logging {
       hadoopConf.get("fs.s3a.endpoint.region") == null) {
       // set to US central endpoint which can also connect to buckets
       // in other regions at the expense of a HEAD request during fs creation
-      hadoopConf.set("fs.s3a.endpoint", "s3.amazonaws.com")
+      hadoopConf.set("fs.s3a.endpoint", "s3.amazonaws.com", setBySpark)
     }
   }
 
   private def appendSparkHiveConfigs(conf: SparkConf, hadoopConf: Configuration): Unit = {
     // Copy any "spark.hive.foo=bar" spark properties into conf as "hive.foo=bar"
     for ((key, value) <- conf.getAll if key.startsWith("spark.hive.")) {
-      hadoopConf.set(key.substring("spark.".length), value)
+      hadoopConf.set(key.substring("spark.".length), value,
+        "Set by Spark from keys starting with 'spark.hive'")
+    }
+  }
+
+  /**
+   * Extract the origin of a configuration key, or a default value if
+   * the key is not found or its origin is not known.
+   * Note that options provided by credential providers (JCEKS stores etc)
+   * are not resolved, so values retrieved by Configuration.getPassword()
+   * may not be recorded as having an origin.
+   * @param hadoopConf hadoop configuration to examine.
+   * @param key key to look up
+   * @param defVal default value
+   * @return the origin of the current entry in the configuration, or the default.
+   */
+  def extractOrigin(hadoopConf: Configuration, key: String, defVal: String): String = {
+    val sources = hadoopConf.getPropertySources(key)
+    if (sources != null && sources.length > 0) {
+      sources(0)
```

Comment:
I'll do the mkstring(); it can only help debugging to know that entire chain. This is what cloudstore does, incidentally. I also plan to improve the s3a credential providers to log at debug their sources as passed in, maybe even some simple crc2 checksum (so you can use crc32 on the command like to verify your secrets are in, but not enough that if someone has the log they can determine a valid set of credentials)